src/sql/src/func.rs

```diff
@@ -1686,46 +1686,50 @@ macro_rules! catalog_name_only {
-                            )
-                                AS {}
-                        FROM
-                            (
-                                SELECT
-                                    unnest(privileges)
-                                FROM
-                                    {}
-                                WHERE
-                                    {}.oid = $2
-                            )
-                                AS user_privs (privilege)
-                            LEFT JOIN mz_roles ON
-                                    mz_internal.mz_aclitem_grantee(privilege) = mz_roles.id
-                        WHERE
-                            mz_internal.mz_aclitem_grantee(privilege) = '{}' OR pg_has_role($1, mz_roles.oid, 'USAGE')
-                    ),
-                    false
-                )
-                END
-            ",
-            $catalog_tbl, $fn_name, $catalog_tbl, $catalog_tbl, RoleId::Public,
-        )
+        {
+            let fn_name = $fn_name;
+            let catalog_tbl = $catalog_tbl;
+            let public_role = RoleId::Public;
+            &format!(
+                "
+                    CASE
+                    -- We need to validate the privileges to return a proper error before anything
+                    -- else.
+                    WHEN NOT mz_internal.mz_validate_privileges($3)
+                    OR $1 IS NULL
+                    OR $2 IS NULL
+                    OR $3 IS NULL
+                    OR $1 NOT IN (SELECT oid FROM mz_roles)
+                    OR $2 NOT IN (SELECT oid FROM {catalog_tbl})
+                    THEN NULL
+                    ELSE COALESCE(
+                        (
+                            SELECT
+                                bool_or(
+                                    mz_internal.mz_acl_item_contains_privilege(privilege, $3)
+                                )
+                                    AS {fn_name}
+                            FROM
+                                (
+                                    SELECT
+                                        unnest(privileges)
+                                    FROM
+                                        {catalog_tbl}
+                                    WHERE
+                                        {catalog_tbl}.oid = $2
+                                )
+                                    AS user_privs (privilege)
+                                LEFT JOIN mz_roles ON
+                                        mz_internal.mz_aclitem_grantee(privilege) = mz_roles.id
+                            WHERE
+                                mz_internal.mz_aclitem_grantee(privilege) = '{public_role}' OR pg_has_role($1, mz_roles.oid, 'USAGE')
+                        ),
+                        false
+                    )
+                    END
+                ",
+            )
+    }
```

Comment:
The only changes here are to put the actual variable name in-between the `{}` to make it easier to read. Hide whitespace for better viewing.